packages/astro/src/core/build/plugins/plugin-content.ts

```diff
@@ -0,0 +1,302 @@
+import { type BuildInternals } from '../internal.js';
+import type { AstroBuildPlugin } from '../plugin.js';
+import type { StaticBuildOptions } from '../types.js';
+import { generateContentEntryFile, generateLookupMap } from '../../../content/vite-plugin-content-virtual-mod.js';
+import { joinPaths, removeFileExtension, removeLeadingForwardSlash } from '../../path.js';
+import { fileURLToPath } from 'node:url';
+import { type ContentLookupMap, hasContentFlag } from '../../../content/utils.js';
+import { CONTENT_RENDER_FLAG, PROPAGATED_ASSET_FLAG } from '../../../content/consts.js';
+import { copyFiles } from '../static-build.js';
+import pLimit from 'p-limit';
+import { extendManualChunks } from './util.js';
+import { isServerLikeOutput } from '../../../prerender/utils.js';
+import { encodeName } from '../util.js';
+
+const CONTENT_CACHE_DIR = './content/';
+const CONTENT_MANIFEST_FILE = './manifest.json';
+// IMPORTANT: Update this version when making significant changes to the manifest format.
+// Only manifests generated with the same version number can be compared.
+const CONTENT_MANIFEST_VERSION = 0;
+
+interface ContentManifestKey {
+	collection: string;
+	type: 'content' | 'data';
+	entry: string;
+}
+interface ContentManifest {
+	version: number;
+	entries: [ContentManifestKey, string][];
+	// Tracks components that should be included in the server build
+	// When the cache is restored, these might no longer be referenced
+	serverEntries: string[];
+	// Tracks components that should be passed to the client build
+	// When the cache is restored, these might no longer be referenced
+	clientEntries: string[];
+}
+
+const virtualEmptyModuleId = `virtual:empty-content`;
+const resolvedVirtualEmptyModuleId = `\0${virtualEmptyModuleId}`;
+
+function createContentManifest(): ContentManifest {
+	return { version: -1, entries: [], serverEntries: [], clientEntries: [] };
+}
+
+function vitePluginContent(opts: StaticBuildOptions, lookupMap: ContentLookupMap, internals: BuildInternals): VitePlugin {
+	const { config } = opts.settings;
+	const { cacheDir } = config;
+	const distRoot = config.outDir;
+	const distContentRoot = new URL('./content/', distRoot);
+	const cachedChunks = new URL('./chunks/', opts.settings.config.cacheDir);
+	const distChunks = new URL('./chunks/', opts.settings.config.outDir);
+	const contentCacheDir = new URL(CONTENT_CACHE_DIR, cacheDir);
+	const contentManifestFile = new URL(CONTENT_MANIFEST_FILE, contentCacheDir);
+	const cache = contentCacheDir;
+	const cacheTmp = new URL('./.tmp/', cache);
+	let oldManifest = createContentManifest();
+	let newManifest = createContentManifest();
+	let entries: ContentEntries;
+	let injectedEmptyFile = false;
+
+	if (fsMod.existsSync(contentManifestFile)) {
+		try {
+			const data = fsMod.readFileSync(contentManifestFile, { encoding: 'utf8' });
+			oldManifest = JSON.parse(data);
+			internals.cachedClientEntries = oldManifest.clientEntries;
```

Comment:
I think it's fine if we suppress the `readFileSync` error, but a `JSON.parse` error can mean that the cache was tampered with, and we should tell it to the user somehow. And we should **wipe** the current cache because it's unusable.